wgengine/magicsock/magicsock.go

```diff
@@ -1100,19 +1100,20 @@ func (a *AddrSet) UpdateDst(new *net.UDPAddr) error {
 	a.mu.Lock()
 	defer a.mu.Unlock()
 
-	if a.roamAddr != nil {
-		if equalUDPAddr(a.roamAddr, new) {
-			// Packet from the current roaming address, no logging.
-			// This is a hot path for established connections.
-			return nil
-		}
-	} else if a.curAddr >= 0 && equalUDPAddr(new, &a.addrs[a.curAddr]) {
+	if a.roamAddr != nil && equalUDPAddr(a.roamAddr, new) {
+		// Packet from the current roaming address, no logging.
+		// This is a hot path for established connections.
+		return nil
+	}
+	if a.curAddr >= 0 && equalUDPAddr(new, &a.addrs[a.curAddr]) {
 		// Packet from current-priority address, no logging.
 		// This is a hot path for established connections.
 		return nil
-	} else if new.IP.Equal(derpMagicIP) {
-		// Never pick DERP addresses as a roaming addr. DERP obeys its
-		// own endpoint selection logic.
+	}
+	if new.IP.Equal(derpMagicIP) {
```

Comment:
If it's a hot path, it can be done before the Mutex.Lock on line 1100. It doesn't depend on roamAddr or anything guarded by the mutex.